contoso_manufacturing/developer/webapp-decode/yolov8.py

```diff
@@ -22,37 +24,64 @@ def __init__(self, rtsp_url, class_names, input_shape, color_palette, confidence
         self.verbose=verbose
         self.frame_number =0
         self.skip_rate=skip_rate
-
+        self.grpc_client = make_grpc_client(ovms_url)
+        self.stopped = False
+        self.lock = threading.Lock()
+        # captures the frames and preprocesses them
+        self.preprocessed_frames_queue = queue.Queue(maxsize=150)
+        # captures the frames and outputs from the inference
+        self.inferenced_frames_queue = queue.Queue(maxsize=150)
+        # postprocesses the frames
+        self.postprocessed_frames_queue = queue.Queue(maxsize=150)
+        self.capture_thread = threading.Thread(target=self.capture_frames)
+        self.capture_thread.start()
+        self.postprocess_thread = threading.Thread(target=self.postprocess_frames)
+        self.postprocess_thread.start()
+        self.inference_thread = threading.Thread(target=self.run_inference)
+        self.inference_thread.start()
         # Track frames and inference processing time for displaying FPS performance metrics 
         self.total_inference_time = 0.0
         self.inference_fps = 0.0
         self.total_fps = 0.0
         self.total_frames = 0
         self.start_time = time.time()
 
-        self.cap = cv2.VideoCapture(rtsp_url)
-        self.grpc_client = make_grpc_client(ovms_url)
-        
-        if not self.cap.isOpened():
-            print("Error: Unable to open video source.")
-        else:
-            # Lee un frame para determinar el tamaño de los frames del video
-            ret, frame = self.cap.read()
-            if ret:
-                self.img_height, self.img_width = frame.shape[:2]
-                print(f"Image dimensions: {self.img_width}x{self.img_height}")
-            else:
-                print("Failed to grab frame to set image dimensions")
+    def capture_frames(self):
+        cap = cv2.VideoCapture(self.rtsp_url)
+        while not self.stopped:
+            ret, frame = cap.read()
+            if not ret:
+                print("Failed to grab frame")
+                break
+            
+            preprocessed_frame = self.preprocess(frame)
+            frame_tuple = (frame, preprocessed_frame)
+            while self.preprocessed_frames_queue.full():
```

Comment:
On delete, is there a risk that this thread (and / or others with the same logic below) could get stuck in an infinite loop where "self.stopped" has been set to true, but the thread is stuck in an infinite 'sleep' loop?

It might be better to change this 'while queue full' loop to an 'if queue full' test and then add a 'continue' after each sleep.  This means that the 'self.stopped' condition will be evaluated after each sleep (avoiding the risk of getting stuck during resource deletion).

Reply:
Thanks for catching this, fixed this in all the loops